doc/style-guide-testing.md

```diff
@@ -0,0 +1,467 @@
+
+
+Data Sets
+---------
+
+A filter is often tested by providing a series of inputs and asserting the output is correct.
+
+```
+/*global beforeEach, describe, expect, inject, it, module*/
+describe('myFilter', function () {
+    var filter;
+
+    beforeEach(module('myModule'));
+    beforeEach(inject(function (myFilter) {
+        filter = myFilter;
+    }));
+
+    // Define an array of scenarios to test
+    [
+        {
+            input: undefined,
+            it: 'converts undefined to empty string',
+            output: ''
+        },
+        {
+            input: 7,
+            it: 'converts 7 to "7"',
+            output: '7'
+        },
+        {
+            input: 'Ian'
+            it: 'adds Ian\'s last name',
+            output: 'Ian Finch'
+        }
+    ].forEach(function (scenario) {
+        it(scenario.it, function () {
+            expect(myFilter(scenario.input)).toEqual(scenario.output);
+        });
+    });
+});
+```
+
+
+Common Tests
+============
+
+Sometimes a directive may call an internal function that performs specific actions.  In these cases you want to ensure those things do exactly what they should do if that internal function is called.
+
+    /*global angular, beforeEach, describe, expect, inject, it*/
+    describe('myDirective', function () {
+        var compile, eventData, eventList;
+
+        function testEffectsOfInternalFunction(scope) {
+            it('emits "goAhead", function () {
+                expect(eventList.indexOf('goAhead')).toBe(2);  // Third event
+                expect(eventList.lastIndexOf('goAhead')).toBe(2);  // Only once
+            });
+            it('set validValue to true', function () {
+                expect(scope.validValue).toBe(true);
+            });
+        }
+
+        function testNegativeEffectsOfInternalFunction(scope) {
+            it('never emitted "goAhead"', function () {
+                expect(eventList.indexOf('goAhead')).toBe(-1);
+            });
+            it('deleted or never set validValue', function () {
+                expect(scope.validValue).toBe(undefined);
+            });
+        }
+
+        beforeEach(module('myApp'));
+        beforeEach(inject(function ($compile, $rootScope) {
+            compile = function (input) {
+                var scope;
+
+                scope = $rootScope.$new();
+                $compile(angular.element('<div my-directive="input"></div>'))(scope)
+
+                // Directive makes its own isolated scope
+                return scope.$$childHead;
+            };
+        }));
+        it('allows numbers', function () {
```

Comment:
For the few tests below it looks like they do the same thing but the descriptions in `it()` seem to suggest it should be testing something different.